cephfs/path_test.go

```diff
@@ -60,3 +60,273 @@ func TestRemoveDir(t *testing.T) {
 	assert.EqualError(t, err,
 		fmt.Sprintf("stat %s: no such file or directory", localPath))
 }
+
+func TestLink(t *testing.T) {
+	mount := fsConnect(t)
+	defer fsDisconnect(t, mount)
+
+	t.Run("directory operations", func(t *testing.T) {
+		// Root dir, both as source and destination.
+		err := mount.Link("/", "/")
+		// Error directory operations are not allowed.
+		assert.Error(t, err)
+
+		dirname := "mydir"
+		localPath := path.Join(CephMountDir, dirname)
+		err = mount.MakeDir(dirname, 0755)
+		assert.NoError(t, err)
+		_, err = os.Stat(localPath)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, mount.RemoveDir(dirname))
+		}()
+
+		// Creating link for a directory.
+		err = mount.Link(dirname, "/")
+		// Error, directory operations not allowed.
+		assert.Error(t, err)
+
+		filename := "anyFile.txt"
+		err = mount.Link(dirname, filename)
+		// Error, directory operations not allowed.
+		assert.Error(t, err)
+
+		err = mount.Link(filename, dirname)
+		// Error, file does not exist.
+		assert.Error(t, err)
+
+		fname := "TestFile.txt"
+		f1, err := mount.Open(fname, os.O_WRONLY|os.O_CREATE, 0666)
+		assert.NotNil(t, f1)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname))
+		}()
+		err = mount.Link(fname, dirname)
+		// Error, destination exists.
+		assert.Error(t, err)
+	})
+
+	t.Run("file operations", func(t *testing.T) {
+		fname := "notExist.txt"
+		err := mount.Link(fname, "hardlnk")
+		// Error, file does not exist.
+		assert.Error(t, err)
+
+		// Hard link creation.
+		fname1 := "TestFile1.txt"
+		f1, err := mount.Open(fname1, os.O_WRONLY|os.O_CREATE, 0666)
+		assert.NotNil(t, f1)
+		assert.NoError(t, err)
+		defer func() {
+			assert.NoError(t, f1.Close())
+			assert.NoError(t, mount.Unlink(fname1))
+		}()
+		err = mount.Link(fname1, "hardlnk")
+		defer func() { assert.NoError(t, mount.Unlink("hardlnk")) }()
+		// No error, normal link operation.
+		assert.NoError(t, err)
+		// Verify that link got created.
+		_, err = os.Stat(path.Join(CephMountDir, "hardlnk"))
+		assert.NoError(t, err)
+
+		// Create hard link when destination exists.
```

Comment:
Another sub-section that should probably be its own sub-test.

Reply:
Done